openfl/federated/data/sources/verifiable_dataset_info.py

```diff
@@ -0,0 +1,146 @@
+from pathlib import Path
+from typing import List
+
+from openfl.federated.data.sources.data_source import DataSource, DataSourceType
+from openfl.federated.data.sources.local_data_source import LocalDataSource
+
+
+class DatasetFormat(Enum):
+    """Enum for the different dataset commitment formats."""
+
+    VERBOSE = "verbose_dataset"
+    CONCISE = "concise_dataset"
+
+
+def filter_non_serializable(obj):
+    """Filter out methods and non-serializable objects."""
+    serializable_dict = {}
+    for key, val in obj.__dict__.items():
+        if callable(val):
+            continue  # Skip methods
+        if isinstance(val, Enum):
+            val = val.value  # Convert Enum to its value
+        elif isinstance(val, Path):
+            val = str(val)  # Convert Path to string
+        serializable_dict[key] = val
+    return serializable_dict
+
+
+class VerifiableDatasetInfo:
+    """
+    This class represents a data set whose integrity can be verified.
+    It contains multiple data sources and methods to compute data commitments, verification,
+    as well as utilities for serialization and deserialization.
+    """
+
+    def __init__(
+        self,
+        data_sources: List[DataSource],
+        base_path: Path,
+        label: str,
+        metadata,
+        root_hash=None,
+    ):
+        self.data_sources = data_sources
+        self.label = label
+        self.metadata = metadata
+        self.base_path = Path(base_path)
+        self.root_hash = root_hash
+        self.all_hashes = None
+        if self.root_hash is None:
+            self.root_hash = self.create_dataset_hash()
+
+    def _create_verbose_dataset_hash(self):
+        self.all_hashes = {
+            str(file_path.relative_to(self.base_path)): ds.compute_object_hash(
+                str(self.base_path / file_path)
+            )
+            for ds in self.data_sources
+            for file_path in ds.enumerate_objects(str(self.base_path))
+        }
+        return self.all_hashes
+
+    def _create_concise_dataset_hash(self):
+        all_file_hashes = self._create_verbose_dataset_hash()
+        sorted_file_hashes = sorted(all_file_hashes.values())
+        joined_hashes = "".join(sorted_file_hashes)
+        self.root_hash = sha384(joined_hashes.encode()).hexdigest()
+        return self.root_hash
+
+    def create_dataset_hash(self):
+        """Create and return the root_hash of all files hashes."""
+        return self._create_concise_dataset_hash()
+
+    def _validate_verbose_dataset_info(self):
+        hashes = self._create_verbose_dataset_hash()
+        return sorted(hashes.values()) == sorted(self.all_hashes.values())
+
+    def _validate_concise_dataset_info(self):
+        concise_hash = self._create_concise_dataset_hash()
+        return concise_hash == self.root_hash
+
+    def verify_dataset(self, dataset_info=None):
+        """Verify the dataset root_hash."""
+        if dataset_info is None and self.root_hash is None:
```

Comment:
What does the `and` check? From the exception it seems the intent is to only check `if not dataset_info: raise ValueError`. Potential for simplification or clarification in the error message.

Reply:
The verification can be against dataset_info or already saved root_hash (for when we don't have the dataset_info anymore). I'll fix the error message for clarification, thanks.